agent/consul/server.go

```diff
@@ -425,6 +404,29 @@ func NewServer(config *Config, flat Deps, publicGRPCServer *grpc.Server) (*Serve
 		fsm:                     newFSMFromConfig(flat.Logger, gc, config),
 	}
 
+	var recorder *middleware.RequestRecorder
+	if flat.NewRequestRecorderFunc != nil {
+		recorder = flat.NewRequestRecorderFunc(serverLogger, s.IsLeader, s.config.Datacenter)
```

Comment:
I'm not sure what are the implication of passing `IsLeader` to the request recorder in here. IsLeader go as deep as atomically reading the raft state to be able to tell and that could have sides effects that are hard to reason about (specially performance effects)
I would recommend to have a separate state for the recorder that we can update when we acquire or loose leadership.

Reply:
> I would recommend to have a separate state for the recorder that we can update when we acquire or loose leadership.

Would you mind chatting thru this some more with me? When I thought about this, I started to think that it can get too complex keeping tabs on the leadership state for a server but I may be wrong here 🤔 

---

On another note, I see this "pattern" (of passing the func/ a smaller interface) being loosely used in `gateway` code here:

https://github.com/hashicorp/consul/blob/09d61e643f4faacb6eb19e2252f8641549d9e7c3/agent/consul/gateway_locator.go#L272-L277

Reply:
We were chatting offline and believe that the underlying `raft.getState()` per RPC call would not adversely impact performance.

We are still interested in coming up with a different way of the server telling the request recorder about its leadership state. I will backlog an issue for this.